libs/utils/android/benchmark.py

```diff
@@ -0,0 +1,287 @@
+            self._log.info('Using ACME energy meter channels: %s', em)
+
+        # Override EM if energy collection not required
+        if 'energy' not in self.bm_collect:
+            try:
+                self.bm_conf.pop('emeter')
+            except:
+                pass
+
+        return self.bm_conf
+
+    def _getBmName(self):
+        if self.bm_name is None:
+            msg = 'Benchmark subclasses must override the `bm_name` attribute'
+            raise NotImplementedError(msg)
+        # Get a referench to the worload to run
+        wl = Workload.getInstance(self.te, self.bm_name)
+        if wl is None:
+            raise ValueError('Specified benchmark [{}] is not supported'\
+                             .format(self.bm_name))
+        return wl
+
+    def _getBmParams(self):
+        if self.bm_params is None:
+            msg = 'Benchmark subclasses must override the `bm_params` attribute'
+            raise NotImplementedError(msg)
+        return self.bm_params
+
+    def _getBmCollect(self):
+        if self.bm_collect is None:
+            msg = 'Benchmark subclasses must override the `bm_collect` attribute'
+            self._log.warning(msg)
+            return ''
+        return self.bm_collect
+
+    def __init__(self):
+        """
+        Set up logging and trigger running experiments
+        """
+        LisaLogging.setup()
+        self._log = logging.getLogger('Benchmark')
+
+        self._log.info('=== CommandLine parsing...')
+        self.args = self._parseCommandLine()
+
+        self._log.info('=== TestEnv setup...')
+        self.bm_conf = self._getBmConf()
+        self.te = TestEnv(self.bm_conf)
+        self.target = self.te.target
+
+        self._log.info('=== Initialization...')
+        self.wl = self._getBmName()
+        self.out_dir=self.te.res_dir
+        try:
+            self.benchmarkInit()
+        except:
+            self._log.warning('Benchmark initialization failed: execution aborted')
+            raise
+
+        self._log.info('=== Execution...')
+        self.wl.run(out_dir=self.out_dir,
+                    collect=self._getBmCollect(),
+                    **self.bm_params)
+
+        self._log.info('=== Finalization...')
+        self.benchmarkFinalize()
+
+    def _wait_for_logcat_idle(self, seconds=1):
+        lines = 0
+
+        # Clear logcat
+        # os.system('{} logcat -s {} -c'.format(adb, DEVICE));
+        self.target.clear_logcat()
+
+        # Dump logcat output
+        logcat_cmd = 'adb -s {} logcat'.format(self.target.adb_name)
+        logcat = Popen(logcat_cmd, shell=True, stdout=PIPE)
+        logcat_poll = select.poll()
+        logcat_poll.register(logcat.stdout, select.POLLIN)
+
+        # Monitor logcat until it's idle for the specified number of [s]
+        self._log.info('Waiting for system to be almost idle')
+        self._log.info('   i.e. at least %d[s] of no logcat messages', seconds)
+        while True:
```

Comment:
I've added a limit on number of processed lines (1e6)... but I forgot to add the break :-/